### ISSUE TITLE
Require that all substitutions used in a project are provided to the image builder

### ISSUE THREAD

**Issue by vjovanov** [oracle/graal#8683]
### TL;DR

The [Native Image substitutions](https://github.com/oracle/graal/blob/master/sdk/src/com.oracle.svm.core.annotate/src/com/oracle/svm/core/annotate/Substitute.java) are currently applied if they are found on the image-builder class path. This is problematic for the following reasons: 
1. Users can not disable incorrect substitutions in third-party projects.
2. The class-path scanning is necessary making the image build longer. Class path scanning needs to be [removed in the future](https://github.com/oracle/graal/issues/2599).

We need to make the usage of all substitutions explicit. This can be done via a new value `Substitutions` in the `native-image.properties` file. The value `Substitutions` would be a coma-separated list of all fully-qualified classes annotated with [`@TargetClass`](https://github.com/oracle/graal/blob/master/sdk/src/com.oracle.svm.core.annotate/src/com/oracle/svm/core/annotate/TargetClass.java). These values would be susceptible to the `--exclude-config` flag which would make it possible for people to disable third-party substitutions.

#### Goals

1. Make it possible to disable erroneous third-party substitutions.
2. Make the class-path scanning unnecessary for the usage substitutions.

#### Non-goals

1. Make abrupt breaking changes the existing code in the community. 
2. Disallow usage of substitutions.

#### Migration

In the first step, the substitutions that are discovered on the class path but that are not specified in the `Substitutions` value would be output in a warning message:
```
Warning: The following substitutions were discovered on the class path but were not specified explicitly in the `native-image.properties` file. Please add the following entries to the class path: 
    'Substitutions= ' in <cp-entry-1>/META-INF/native-image/<artifact-id>/native-image.properties
...
    'Substitutions= ' in <cp-entry-n>/META-INF/native-image/<artifact-id>/native-image.properties

In case these files are not under your control please open a ticket for the library authors or submit a PR to the metadata repository.
```

In the next step we would convert the warning into an error message with a possibility to disable the error via a flag.

In the third step there would not even be an error anymore, but the substitutions would simply be ignored.

The time duration between steps 1, 2, and 3 is to be discussed as it depends on the number of substitutions used in the community.

#### Open Questions

1. Should we allow to specify the substitutions list in the [metadata repository](https://github.com/oracle/graalvm-reachability-metadata)?
2. Should we have a flag for specifying substitutions? API flag or not?
3. Should we provide an annotations processor for the substitutions that will populate the `native-image.properties` automatically? How would it know in which folder on the class path to store the values?

**Comment by turing85** [comment]
Can there be cases where users want to exclude specific substitutions instead of all substitutions within a `native-image.properties`? Should we have a possibility to exclude single substitutions within a `native-image.properties`?

**Comment by vjovanov** [comment]
@turing85 this is a good question. This also noted by @fniephaus. I think such flag would not hurt as long as it is not an API flag. I will update the writing to add such flag.

**Comment by zakkak** [comment]
> 1. Make it possible to disable erroneous third-party substitutions.

This is implemented in https://github.com/oracle/graal/pull/9743 (not merged yet)
